### PR TITLE
bib: Add root filesystem selection (HMS-3110)

### PR DESCRIPTION
**Pull request by bcl**
Add `--rootfs` to select the root filesystem type (eg. ext4, xfs). It defaults to ext4.

### CODE REVIEW

**Comment by bcl** [comment]
I think this is complete, but when I try running things with a locally built container I am getting errors from seemingly unrelated things. When I look at the generated manifest it looks ok, with xfs instead of ext4 for the `/` filesystem type.

```
mount/ostree-ostree/1/1/0 (org.osbuild.ostree.deployment): Deployment root at 'ostree/deploy/default/deploy/719b26bb969e44e8f88a5874c8a519d67ed422a6c0ad06b69991fda3177ff620.0'
[sss_cache] [confdb_init] (0x0010): Unable to open config database [/var/lib/sss/db/config.ldb]          
Could not open available domains                                                                         
[sss_cache] [confdb_init] (0x0010): Unable to open config database [/var/lib/sss/db/config.ldb]          
Could not open available domains                                                                         
Traceback (most recent call last):                                                                       
  File "/run/osbuild/bin/org.osbuild.users", line 199, in <module>                                       
    r = main(args["tree"], args["options"])                                                              
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                              
  File "/run/osbuild/bin/org.osbuild.users", line 182, in main                                           
    ensure_homedir(tree, name, home)                                                                     
  File "/run/osbuild/bin/org.osbuild.users", line 157, in ensure_homedir                                 
    subprocess.run(["chroot", root, "mkhomedir_helper", name], check=True)                               
  File "/usr/lib64/python3.12/subprocess.py", line 571, in run                                           
    raise CalledProcessError(retcode, process.args,  
subprocess.CalledProcessError: Command '['chroot', '/run/osbuild/tree', 'mkhomedir_helper', 'root']' returned non-zero exit status 6.
```

**Comment by mvo5** [comment]
Thanks, this looks very nice (and exciting!). 

I wonder if you also see these unexpected errors if you add the following:
```python
diff --git a/test/test_smoke.py b/test/test_smoke.py
index 03e9fd8..497629e 100644
--- a/test/test_smoke.py
+++ b/test/test_smoke.py
@@ -40,7 +40,7 @@ def build_container_fixture():
 
 
 # image types to test
-SUPPORTED_IMAGE_TYPES = ["qcow2", "ami"]
+SUPPORTED_IMAGE_TYPES = ["qcow2,xfs", "qcow2,btrfs", "ami,ext4"]
 
 
 class ImageBuildResult(NamedTuple):
@@ -48,6 +48,7 @@ class ImageBuildResult(NamedTuple):
     username: str
     password: str
     journal_output: str
+    rootfs: str
 
 
 @pytest.fixture(name="image_type", scope="session")
@@ -64,7 +65,7 @@ def image_type_fixture(tmpdir_factory, build_container, request):
     )
 
     # image_type is passed via special pytest parameter fixture
-    image_type = request.param
+    image_type, rootfs = request.param.split(",")
 
     username = "test"
     password = "password"
@@ -116,11 +117,12 @@ def image_type_fixture(tmpdir_factory, build_container, request):
         container_to_build_ref,
         "--config", "/output/config.json",
         "--type", image_type,
+        "--rootfs", rootfs,
     ])
     journal_output = testutil.journal_after_cursor(cursor)
     journal_log_path.write_text(journal_output, encoding="utf8")
 
-    return ImageBuildResult(generated_img, username, password, journal_output)
+    return ImageBuildResult(generated_img, username, password, journal_output, rootfs)
 
 
 def test_container_builds(build_container):
```
and then run:
```
$ sudo pytest -s -vv 'test/test_smoke.py::test_image_boots[qcow2,xfs]'
```
this seems to work for me. Adding code that the fs is correct could be added via:
```python
diff --git a/test/test_smoke.py b/test/test_smoke.py
index 03e9fd8..2c48cfe 100644
--- a/test/test_smoke.py
+++ b/test/test_smoke.py
@@ -144,6 +146,9 @@ def test_image_boots(image_type):
         exit_status, output = test_vm.run("echo hello", user=image_type.username, password=image_type.password)
         assert exit_status == 0
         assert "hello" in output
+        exit_status, output = test_vm.run("findmnt -no FSTYPE -T /", user=image_type.username, password=image_type.password)
+        assert exit_status == 0
+        assert output.strip() == image_type.rootfs
 
 
 def log_has_osbuild_selinux_denials(log):
```
(on top of the above). And that also works for "xfs" which is really nice.

What is interesting is that xfs seems to be fine but btrfs fails with:
```
...
Generating manifest-qcow2.json ... No match for argument: btrfs-progs
error install_specs
MarkingErrors: Error occurred when marking packages for installation: Problems in request:
missing packages: btrfs-progs
DNF error occurred: MarkingErrors: Error occurred when marking packages for installation: Problems in request:
missing packages: btrfs-progs
ERROR
```
I add the full diff that I used for completeness 
[0001-test-add-support-to-test-images-with-configurable-ro.patch.txt](https://github.com/osbuild/bootc-image-builder/files/13901752/0001-test-add-support-to-test-images-with-configurable-ro.patch.txt)

**Comment by cgwalters** [comment]
> Add --rootfs to select the root filesystem type (eg. ext4, xfs). It defaults to ext4.

Note this overlaps with https://github.com/containers/bootc/blob/main/docs/install.md#operating-system-install-configuration-required

I'd really like to see us fix https://github.com/osbuild/bootc-image-builder/issues/18 because we would then already have filesystem selection instead of re-implementing it.

**Review by cgwalters — DISMISSED** [review]
Per above I think we should target bootc functionality.

**Comment by bcl** [comment]
> Per above I think we should target bootc functionality.

sure, but this is my effort to knock out some low hanging fruit, and to get more familiar with this codebase. So until there is actual bootc work being done I don't see any reason to block this. It's pretty simple.

**Comment by bcl** [comment]
> I wonder if you also see these unexpected errors if you add the following:

Actually no, that seems to work fine. I'm seeing them with my own run script and config.json, so I may be doing something wrong with them...

**Comment by cgwalters** [comment]
> sure, but this is my effort to knock out some low hanging fruit, and to get more familiar with this codebase. 

But the high level goal of the effort branded "bootc" is to make container images the technical heart of things.  This project is now part of that.  Of course, I think everyone is very well aware that we are reusing a giant amount of prior existing code (for example, Anaconda) because it just makes sense.

We have already set a precedent for external-to-container configuration being provided in this project - and that was agreed because as a short term expedience thing managing credentials and secrets is something one may often want to ship outside of the container image (disk images can be managed externally, i.e. secrets may just be one-time provisioning state, as users provided are).

At the same time, we should also aim to have the focus be on containers as where the operating system is configured.  And in *this specific* case, we already have functionality in bootc to do this.  Adding more things like this sets increasing precedent for managing OS state and source of truth outside of containers, which is misaligned with the overall goal.

**Comment by rhatdan** [comment]
If bootc can do this now, why not use it?

**Comment by cgwalters** [comment]
I'd also emphasize that I do feel to some degree like the `bootc install` is still subject to change - and everyone working on this if you have opinions on how it could do things differently please do speak up!  You should all feel as if you have a say in its future.

**Comment by cgwalters** [comment]
This all said, it should absolutely be supported to override the default OS filesystem as a downstream user of an externally-produced base image.

The classic situation here is wanting to e.g. use xfs/bcachefs/etc as a Fedora Silverblue user or whatever - decoupling the ownership of base image from filesystem choice.

Our current primary target for bootc though is where user-owns-OS: the expected entrypoint to the system is deriving from our base image.  And now today, we have an overlap between:

- configuration via container
- configuration via image

That said, I still do also want to enable "generic" image use even in the bootc sphere.  This is one reason we have https://github.com/CentOS/centos-bootc-layered/tree/main/cloud which basically acts as a very compatible low-cost entrypoint to the system - you can boot it and get a root shell via ssh etc. that works the same as many distribution "generic cloud" images *without* building your own container.

In the same vein as this PR, we already have explicit support for downstream filesystem customization via kickstart.

So I guess, I will just lift my changes-requested.

But at the same time, I hope we can slowly start to reconcile these things (kickstart, "bib config", blueprints, and finally bootc install configs).

File: README.md [hunk 1]
```diff
@@ -117,6 +117,7 @@ Flags:
 | **--config** | Path to a [build config](#-build-config)                         |       ❌      |
 | --tls-verify | Require HTTPS and verify certificates when contacting registries |    `true`     |
 | **--type**   | [Image type](#-image-types) to build                             |    `qcow2`    |
+| --rootfs     | Root filesystem type to create. eg. ext4, xfs                    |    `ext4`     |
```
Comment (cgwalters) [inline README.md:120]: Can you at least just document here that this functionality will also be accessible in the future via `bootc install` configs?

Comment (bcl) [reply README.md:120]: Isn't that getting ahead of ourselves though? I think it's confusing to refer to future things that don't work right now.

**Comment by bcl** [comment]
Thanks for the additional comments. Where I'm getting lost is the mention of all the other possible configuration options (most of which aren't yet available?). eg. how does Anaconda and Kickstart fit into `bib` for example? AFAIK this tool is making ami or qcow2 images for booting on bare metal so there is no kickstart.

I agree that it's better to have one config than a pile of them :) I'm just not sure what that is yet (and would lean towards blueprints which are already part of the `bib` configuration).

**Pull request closed without merging**